File: TDDMicroExercises.ProposedSolution/CSharp/TelemetrySystem/TelemetryDiagnosticControls.cs
using System;

namespace TDDMicroExercises.OneSolution.TelemetrySystem
{
    public class TelemetryDiagnosticControls
    {
        private const string DiagnosticChannelConnectionString = "*111#";
        
        private readonly IConnection _connection;
        private readonly IDataChannel _dataChannel;

        private string _diagnosticInfo = string.Empty;


        public TelemetryDiagnosticControls(IConnection connection, IDataChannel dataChannel)
        {
            _connection = connection;
            _dataChannel = dataChannel;
        }


        public TelemetryDiagnosticControls()
            : this(new TelemetryClientConnection(), new TelemetryDataChannel())
        {
        }


        public string DiagnosticInfo
        {
            get { return _diagnosticInfo; }
            set { _diagnosticInfo = value; }
        }

        public void CheckTransmission()
        {
            _diagnosticInfo = string.Empty;

            _connection.Disconnect();

            int retryLeft = 3;
            while (_connection.OnlineStatus == false && retryLeft > 0)
            {
                _connection.Connect(DiagnosticChannelConnectionString);
                retryLeft -= 1;
            }

            if (_connection.OnlineStatus == false)
            {
                throw new Exception("Unable to connect.");
            }

            _dataChannel.Send(TelemetryDataChannel.DiagnosticMessage);
            _diagnosticInfo = _dataChannel.Receive();
            _connection.Disconnect();
        }
    }
}

File: TDDMicroExercises.ProposedSolution/Java/TelemetrySystem/src/main/java/tddmicroexercises/telemetrysystem/TelemetryDiagnosticControls.java
package tddmicroexercises.telemetrysystem;

import java.net.ConnectException;

public class TelemetryDiagnosticControls
{
    public static final String CONNECTION_STRING = "*111#";
    
    private final IConnection connection;
    private final IDataChannel dataChannel;
    private String diagnosticInfo = "";

        public TelemetryDiagnosticControls(){
            this(new TelemetryConnection(), new TelemetryDataChannel());
        }
        
        public TelemetryDiagnosticControls(IConnection connection, IDataChannel dataChannel) {
			this.connection = connection;
			this.dataChannel = dataChannel;
		}

		public String getDiagnosticInfo(){
        	return diagnosticInfo;
        }
        
        public void setDiagnosticInfo(String diagnosticInfo){
        	this.diagnosticInfo = diagnosticInfo;
        }
 
        public void checkTransmission() throws Exception{
            diagnosticInfo = "";

	        connection.disconnect();
	
	        int retryLeft = 3;
	        while (connection.getOnlineStatus() == false && retryLeft > 0){
	            connection.connect(CONNECTION_STRING);
	            retryLeft -= 1;
	        }
	         
	        if(connection.getOnlineStatus() == false){
	            throw new ConnectException("Unable to connect.");
	        }
	
	        dataChannel.send(IDataChannel.DIAGNOSTIC_MESSAGE);
	        diagnosticInfo = dataChannel.receive();
    }
}
